keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,203 @@
+kep-number: 27
+short-desc: Detailed control what parameter changes trigger pod restarts
+title: Pod restart control
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-03-31
+last-updated: 2020-04-15
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes a feature to allow detailed control when templated pods in a stateful set are restarted when parameters
+that are not used in the template change.
+
+## Motivation
+
+KUDO currently has a feature that automatically restarts Pods of Deployments or StatefulSets if any KUDO-Parameter is changed. This
+is often required as a Parameter can change a value in a ConfigMap that is used by the Pods. If the Pods are not restarted
+the change is propagated to the Pods, but as most applications are not aware of changing config files they require a 
+restart. This was originally solved by using [stakater/Reloader](https://github.com/stakater/Reloader) but this had 
+[issues](https://github.com/kudobuilder/kudo/issues/1025) which lead to the self developed solution, which is at the 
+moment applied to [all templates](https://github.com/kudobuilder/kudo/issues/1036)
+
+This behavior can currently not be controlled and is always active. The issue is that not all Parameter changes require
+a rolling restart of Pods, and sometimes the restart is unwanted and can negatively affect the state of the application
+the operator is controlling.
+
+One example would be the `replica` count of a StatefulSet: An increase here should only start new Pods and not restart
+all existing ones. Another example would be an application that regularly re-reads the config files, in this case the 
+pod doesn't need to be restarted if a config map changes - the application would pick up changed values automatically.
+
+A new flag introduced by this KEP will allow an operator developer to control when a changed parameter will trigger a 
+restart of pods from a deployment or stateful set.
+
+### Goals
+
+Make it possible for an operator developer to specify when a parameter will restart all pods from a stateful set or a deployment.
+
+### Non-Goals
+
+- Preventing Pod restart where plain k8s would do it: If the spec template for a stateful set is changed, Kubernetes will still
+restart all pods automatically - there is no way to prevent that.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+- Restarting a stuck Pod
+
+## Proposal
+
+Allow the operator developer to define groups of parameters that can then be used to trigger a restart of pods from
+a deployment or stateful set.
+
+Add a list of `groups` that each parameter is part of:
```

Comment:
so with this operator developers will have to define which plan each parameter might trigger + define for all the groups the parameter belongs?

Reply:
In the easiest case, to replicate the current behavior, the operator developer would just have to add:
```
spec:
  template:
    metadata:
      annotations:
        config-values-hash: {{ .ParamGroupHashes.ALL }}
```
without defining any custom groups.
But to define custom control, yes, the operator developer would have to add all required parameters to a custom group.

It doesn't affect which plans a parameter is used in though.